File: managed-identity/agent-web-app/src/response-request.js
const MAX_APPROVAL_RESPONSES = 20;

export class ResponseRequestValidationError extends Error {}

export function buildResponseRequest(body) {
  const input = body?.input;
  const previousResponseId = body?.previous_response_id;

  if (
    previousResponseId !== undefined &&
    (typeof previousResponseId !== "string" || previousResponseId.trim().length === 0)
  ) {
    throw new ResponseRequestValidationError(
      "previous_response_id must be a non-empty string.",
    );
  }

  if (typeof input === "string") {
    const message = input.trim();
    if (!message) {
      throw new ResponseRequestValidationError(
        "A non-empty input string is required.",
      );
    }

    return {
      input: message,
      ...(previousResponseId
        ? { previous_response_id: previousResponseId.trim() }
        : {}),
    };
  }

  if (!Array.isArray(input) || input.length === 0) {
    throw new ResponseRequestValidationError(
      "Input must be a non-empty message or MCP approval response array.",
    );
  }

  if (!previousResponseId) {
    throw new ResponseRequestValidationError(
      "previous_response_id is required for MCP approval responses.",
    );
  }

  if (input.length > MAX_APPROVAL_RESPONSES) {
    throw new ResponseRequestValidationError(
      `No more than ${MAX_APPROVAL_RESPONSES} MCP approval responses are allowed.`,
    );
  }

  const approvals = input.map((item) => {
    if (
      !item ||
      typeof item !== "object" ||
      Array.isArray(item) ||
      item.type !== "mcp_approval_response" ||
      typeof item.approval_request_id !== "string" ||
      item.approval_request_id.trim().length === 0 ||
      typeof item.approve !== "boolean"
    ) {
      throw new ResponseRequestValidationError(
        "Each approval must include type mcp_approval_response, a non-empty approval_request_id, and a Boolean approve value.",
      );
    }

    return {
      type: "mcp_approval_response",
      approval_request_id: item.approval_request_id.trim(),
      approve: item.approve,
    };
  });

  return {
    input: approvals,
    previous_response_id: previousResponseId.trim(),
  };
}

File: user-passthrough/agent-web-app/app/response_request.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

MAX_APPROVAL_RESPONSES = 20


class McpApprovalResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["mcp_approval_response"]
    approval_request_id: str = Field(min_length=1)
    approve: bool

    @field_validator("approval_request_id")
    @classmethod
    def strip_approval_request_id(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("approval_request_id must not be empty")
        return stripped


class AgentRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    input: str | list[McpApprovalResponse]
    previous_response_id: str | None = None

    @field_validator("previous_response_id")
    @classmethod
    def strip_previous_response_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        if not stripped:
            raise ValueError("previous_response_id must not be empty")
        return stripped

    @model_validator(mode="after")
    def validate_input(self) -> "AgentRequest":
        if isinstance(self.input, str):
            self.input = self.input.strip()
            if not self.input:
                raise ValueError("input must not be empty")
            return self

        if not self.input:
            raise ValueError("at least one MCP approval response is required")
        if len(self.input) > MAX_APPROVAL_RESPONSES:
            raise ValueError(
                f"no more than {MAX_APPROVAL_RESPONSES} approval responses are allowed"
            )
        if not self.previous_response_id:
            raise ValueError(
                "previous_response_id is required for MCP approval responses"
            )
        return self
